### PR TITLE
Wear os authentication

### PR DESCRIPTION
**Pull request by leroyboerefijn**
## Summary

In this PR a proof-of-concept for authentication on wear os devices. Note that a few changes needed to be made compared to the android app authentication:

- Since resources are very limited on wear OS devices, I decided to omit LaunchActivity. Instead, the checks are done in HomeActivity. This should speed up the launch of HomeActivity (the main flow) slightly
- Since Wear OS devices do not support WebViews, it was not possible to use the default authentication web page and api. Instead, I directly tap into auth/login_flow
- If an android device is connected, it will automatically share the url of the HA instance. This saves the user a lot of annoying typing. It is always possible for the user to manually enter an url if needed.

On HomeActivity, a proof of connection/authentication is given, by loading the home assistant version and the number of instances from the server.

## Screenshots

onboarding
login
integration
home

## Link to pull request in Documentation repository

Documentation: home-assistant/companion.home-assistant#

## Any other notes

Once the authentication is implemented, next steps will be adding functionality to the wear OS app:

- Activating scenes
- Providing sensor information
- Controlling entities

### CODE REVIEW

**Comment by leroyboerefijn** [comment]
@JBassett do you think you could have a look at this PR?  Please let me know if it is too big and needs splitting!

**Comment by hkusulja** [comment]
Can this be the same Android app, with same one on the store, with app manifest for wear os also.
In case it goes to internet through bluetooth and phone, it could share / use same app config (url's, authentication etc.) ?

**Comment by dshokouhi** [comment]
> Can this be the same Android app, with same one on the store, with app manifest for wear os also.
> In case it goes to internet through bluetooth and phone, it could share / use same app config (url's, authentication etc.) ?

you might want to familiarize yourself with how  a standalone app works

https://developer.android.com/training/wearables/overlays/independent-vs-dependent#identify

data can be still be sent from phone to app but the intention is that everything you can do on the phone can also be done on the watch without needing the phone.

**Review by JBassett — CHANGES_REQUESTED** [review]

File: common/src/main/java/io/homeassistant/companion/android/common/data/integration/impl/IntegrationRepositoryImpl.kt [hunk 1]
```diff
@@ -124,6 +124,7 @@ class IntegrationRepositoryImpl @Inject constructor(
     }
 
     override suspend fun isRegistered(): Boolean {
+        Log.d(TAG, urlRepository.getApiUrls().toString())
```
[the remark this answers is not in the document]
Comment (leroyboerefijn) [reply common/src/main/java/io/homeassistant/companion/android/common/data/integration/impl/IntegrationRepositoryImpl.kt]: Agree, I used this for debugging. Have removed it now

File: wear/build.gradle.kts [hunk 2]
```diff
@@ -53,6 +53,25 @@ android {
...
+            versionNameSuffix = "-full"
+        }
+
+        // Generate a list of application ids into BuildConfig
+        val values = productFlavors.joinToString {
+            "\"${it.applicationId ?: defaultConfig.applicationId}${it.applicationIdSuffix}\""
+        }
+
+        defaultConfig.buildConfigField("String[]", "APPLICATION_IDS", "{$values}")
+    }
```
[the remark this answers is not in the document]
Comment (leroyboerefijn) [reply wear/build.gradle.kts]: You're right indeed, removed product flavors

File: wear/src/main/java/io/homeassistant/companion/android/onboarding/viewHolders/ManualSetupViewHolder.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+class ManualSetupViewHolder(v: View, val onClick: () -> Unit) :
+    RecyclerView.ViewHolder(v) {
+
+    val text: TextView = v.findViewById(R.id.name)
+
+    init {
+        // Increase margin
+        val marginTop = (20 * Resources.getSystem().displayMetrics.density).toInt()
+        (v.layoutParams as ViewGroup.MarginLayoutParams).topMargin = marginTop
```
[the remark this answers is not in the document]
Comment (leroyboerefijn) [reply wear/src/main/java/io/homeassistant/companion/android/onboarding/viewHolders/ManualSetupViewHolder.kt]: This does not look great indeed, I removed the margin and added an additional header. This makes it more clear that this is a distinct option

**Comment by leroyboerefijn** [comment]
I don't understand how this error suddenly occurred during the test:

ManualSetupPresenterImplSpec > presenter > on click ok with valid url > io.homeassistant.companion.android.onboarding.manual.ManualSetupPresenterImplSpec.should save the url FAILED
    java.lang.AssertionError at ManualSetupPresenterImplSpec.kt:39
        Caused by: java.lang.AssertionError at ManualSetupPresenterImplSpec.kt:39

**Comment by Chaphasilor** [comment]
hey @leroyboerefijn did you figure out what caused the error? :)

> I don't understand how this error suddenly occurred during the test

I'd love to see this PR get merged, so that actual feature-development could begin! :D

**Comment by dshokouhi** [comment]
The failure is a random test failure, its safe to ignore as every PR is hitting it.  As of now this PR is waiting for another round of review to continue.

**Review by JBassett — APPROVED** [review]

**Comment by leroyboerefijn** [comment]
Thanks for figuring it out @dshokouhi , I was on a holiday!

**Pull request merged**